File: tests/batch_test.cpp
#include "../include/SPSCQueue.h"
#include <iostream>
#include <vector>
#include <cassert>
#include <thread>
#include <cstring>
#include "../src/Platform.h" // For CreateNamedEvent

using namespace shm;

void test_batch_basic() {
    std::cout << "Testing Batch Basic..." << std::endl;
    // 1MB buffer
    size_t shmSize = SPSCQueue::GetRequiredSize(1024 * 1024);
    void* buffer = malloc(shmSize);
    SPSCQueue::Init(buffer, 1024 * 1024);

    // Create real event
    EventHandle hEvent = Platform::CreateNamedEvent("test_batch_basic");

    SPSCQueue queue(buffer, 1024 * 1024, hEvent);

    std::vector<std::vector<uint8_t>> inputs;
    for(int i=0; i<10; ++i) {
        std::vector<uint8_t> msg(100);
        memset(msg.data(), i, 100);
        inputs.push_back(msg);
    }

    // Writer
    queue.EnqueueBatch(inputs);

    // Reader
    std::vector<std::vector<uint8_t>> outputs;
    queue.DequeueBatch(outputs, 100);

    assert(outputs.size() == 10);
    for(int i=0; i<10; ++i) {
        assert(outputs[i].size() == 100);
        assert(outputs[i][0] == i);
    }

    Platform::CloseEvent(hEvent);
    free(buffer);
    std::cout << "PASS" << std::endl;
}

void test_wrapping() {
    std::cout << "Testing Wrapping..." << std::endl;
    // Small buffer: 128 header + 100 capacity
    size_t cap = 200;
    size_t shmSize = SPSCQueue::GetRequiredSize(cap);
    void* buffer = malloc(shmSize);
    SPSCQueue::Init(buffer, cap);

    EventHandle hEvent = Platform::CreateNamedEvent("test_wrapping");
    SPSCQueue queue(buffer, cap, hEvent);

    // Fill slightly
    std::vector<uint8_t> d1(50, 1);
    queue.Enqueue(d1.data(), 50);

    // Read it
    std::vector<uint8_t> out;
    queue.Dequeue(out);
    assert(out.size() == 50);

    // Now wPos is advanced. 50 + 8 = 58.
    // Capacity 200.
    // Let's write enough to wrap.
    // Free space at end: 200 - 58 = 142.
    // Write 100 bytes (total 108).
    std::vector<uint8_t> d2(100, 2);
    queue.Enqueue(d2.data(), 100);

    // wPos: 58 + 108 = 166.
    // Free space at end: 200 - 166 = 34.

    // Write batch that forces wrap.
    // Need > 34 bytes (header+data).
    // Try writing 40 bytes data (+8 = 48).
    // Should pad 34 bytes, wrap to 0, write 48 bytes.
    std::vector<std::vector<uint8_t>> batch;
    batch.push_back(std::vector<uint8_t>(40, 3));

    queue.EnqueueBatch(batch);

    // Read d2
    queue.Dequeue(out);
    assert(out.size() == 100);
    assert(out[0] == 2);

    // Read batch
    std::vector<std::vector<uint8_t>> batchOut;
    queue.DequeueBatch(batchOut, 10);

    assert(batchOut.size() == 1);
    assert(batchOut[0].size() == 40);
    assert(batchOut[0][0] == 3);

    Platform::CloseEvent(hEvent);
    free(buffer);
    std::cout << "PASS" << std::endl;
}

int main() {
    test_batch_basic();
    test_wrapping();
    return 0;
}

File: go/batch_test.go
package shm

import (
	"bytes"
	"testing"
	"unsafe"
)

func TestBatch(t *testing.T) {
	// 1MB
	capacity := uint64(1024 * 1024)
	shmSize := QueueHeaderSize + capacity
	buffer := make([]byte, shmSize)
	base := uintptr(unsafe.Pointer(&buffer[0]))

    // Create real event
    hEvent, err := CreateEvent("test_go_batch")
    if err != nil {
        t.Fatalf("Failed to create event: %v", err)
    }
    defer CloseEvent(hEvent)

	q := NewSPSCQueue(base, capacity, hEvent)
	q.Header.Capacity = capacity

	inputs := make([][]byte, 10)
	for i := range inputs {
		inputs[i] = make([]byte, 100)
		inputs[i][0] = byte(i)
	}

	q.EnqueueBatch(inputs)

	outputs := q.DequeueBatch(100)

	if len(outputs) != 10 {
		t.Fatalf("Expected 10 items, got %d", len(outputs))
	}

	for i, msgPtr := range outputs {
		msg := *msgPtr
		if len(msg) != 100 {
			t.Errorf("Item %d len mismatch", i)
		}
		if msg[0] != byte(i) {
			t.Errorf("Item %d content mismatch", i)
		}
	}

	// Recycle
	q.Recycle(outputs)
}

func TestWrapping(t *testing.T) {
	capacity := uint64(200)
	shmSize := QueueHeaderSize + capacity
	buffer := make([]byte, shmSize)
	base := uintptr(unsafe.Pointer(&buffer[0]))

    // Create real event
    hEvent, err := CreateEvent("test_go_wrapping")
    if err != nil {
        t.Fatalf("Failed to create event: %v", err)
    }
    defer CloseEvent(hEvent)

	q := NewSPSCQueue(base, capacity, hEvent)
	q.Header.Capacity = capacity

	// 1. Write 50 bytes (Total 50 + 16 = 66)
	d1 := make([]byte, 50)
	d1[0] = 1
	q.Enqueue(d1, 0)

	// 2. Read 50
	out, _ := q.Dequeue()
	if len(out) != 50 {
		t.Fatal("Read mismatch")
	}

	// wPos=66, rPos=66. Cap=200.

	// 3. Write 100 bytes (Total 100 + 16 = 116) -> wPos=182
	d2 := make([]byte, 100)
	d2[0] = 2
	q.Enqueue(d2, 0)

	// Space at end: 200-182 = 18.
	// 4. Batch write 40 bytes (Total 40 + 16 = 56). Fits? No, 56 > 18.
	// Should Pad 18 (Header=16, Size=2), wrap, write 56 at 0.

	batch := [][]byte{make([]byte, 40)}
	batch[0][0] = 3
	q.EnqueueBatch(batch)

	// Read d2
	out2, _ := q.Dequeue()
	if !bytes.Equal(out2, d2) {
		t.Fatal("d2 mismatch")
	}

	// Read batch
	batchOut := q.DequeueBatch(10)
	if len(batchOut) != 1 {
		t.Fatal("Batch empty")
	}
	if !bytes.Equal(*batchOut[0], batch[0]) {
		t.Fatal("Batch content mismatch")
	}
}

func TestMsgId(t *testing.T) {
	capacity := uint64(1024)
	shmSize := QueueHeaderSize + capacity
	buffer := make([]byte, shmSize)
	base := uintptr(unsafe.Pointer(&buffer[0]))

    // Create real event
    hEvent, err := CreateEvent("test_go_msgid")
    if err != nil {
        t.Fatalf("Failed to create event: %v", err)
    }
    defer CloseEvent(hEvent)

	q := NewSPSCQueue(base, capacity, hEvent)
	q.Header.Capacity = capacity

	// 1. Write Normal
	d1 := make([]byte, 10)
	d1[0] = 1
	q.Enqueue(d1, MsgIdNormal)

	// 2. Write Heartbeat
	q.Enqueue(nil, MsgIdHeartbeatReq)

	// 3. Write Data with MsgId 2
	d2 := make([]byte, 10)
	d2[0] = 2
	q.Enqueue(d2, 2)

	// Read 1
	out, msgId := q.Dequeue()
	if msgId != MsgIdNormal {
		t.Errorf("Expected MsgIdNormal, got %d", msgId)
	}
	if len(out) != 10 || out[0] != 1 {
		t.Errorf("Data 1 mismatch")
	}

	// Read 2
	out, msgId = q.Dequeue()
	if msgId != MsgIdHeartbeatReq {
		t.Errorf("Expected MsgIdHeartbeatReq, got %d", msgId)
	}
	if len(out) != 0 {
		t.Errorf("Expected empty payload for heartbeat")
	}

	// Read 3
	out, msgId = q.Dequeue()
	if msgId != 2 {
		t.Errorf("Expected MsgId 2, got %d", msgId)
	}
	if len(out) != 10 || out[0] != 2 {
		t.Errorf("Data 2 mismatch")
	}
}
